actors/builtin/miner/miner_actor.go

```diff
@@ -579,6 +628,150 @@ func (a Actor) DeclareTemporaryFaults(rt Runtime, params *DeclareTemporaryFaults
 	return nil
 }
 
+///////////////////////
+// Pledge Collateral //
+///////////////////////
+
+func (a Actor) AwardReward(rt Runtime, _ *adt.EmptyValue) *adt.EmptyValue {
+	rt.ValidateImmediateCallerIs(builtin.RewardActorAddr)
+	pledgeAmount := rt.Message().ValueReceived()
+	Assert(pledgeAmount.GreaterThanEqual(big.Zero()))
+
+	var st State
+	rt.State().Transaction(&st, func() interface{} {
+		if err := st.addLockedFund(adt.AsStore(rt), rt.CurrEpoch(), pledgeAmount); err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to add pledge: %v", err)
+		}
+		return nil
+	})
+
+	a.updateTotalPledge(rt, &pledgeAmount)
+	return nil
+}
+
+func (a Actor) slashPledgeCollateral(rt Runtime, amountToSlash abi.TokenAmount) *abi.TokenAmount {
+
+	if amountToSlash.LessThanEqual(big.Zero()) {
+		rt.Abortf(exitcode.ErrIllegalArgument, "non positive amount to slash: %w", amountToSlash)
+	}
+
+	var st State
+	amountSlashed := rt.State().Transaction(&st, func() interface{} {
+		slashedPledge, err := st.slashLockedFund(adt.AsStore(rt), rt.CurrEpoch(), amountToSlash)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to slash pledge collateral: %v", err)
+		}
+		return slashedPledge
+	}).(abi.TokenAmount)
+
+	_, code := rt.Send(builtin.BurntFundsActorAddr, builtin.MethodSend, nil, amountSlashed)
+	builtin.RequireSuccess(rt, code, "failed to burn funds")
+
+	totalPledgeUpdate := amountSlashed.Neg()
+	a.updateTotalPledge(rt, &totalPledgeUpdate)
+
+	return &amountSlashed
+}
+
+type ReportConsensusFaultParams struct {
+	BlockHeader1     []byte
+	BlockHeader2     []byte
+	BlockHeaderExtra []byte
+}
+
+func (a Actor) ReportConsensusFault(rt Runtime, params *ReportConsensusFaultParams) *adt.EmptyValue {
+
+	// only another miner actor can report consensus fault
+	rt.ValidateImmediateCallerType(builtin.StorageMinerActorCodeID)
+
+	currEpoch := rt.CurrEpoch()
+	earliest := currEpoch - ConsensusFaultReportingWindow
+
+	fault, err := rt.Syscalls().VerifyConsensusFault(params.BlockHeader1, params.BlockHeader2, params.BlockHeaderExtra, earliest)
+	if err != nil {
+		rt.Abortf(exitcode.ErrIllegalArgument, "fault not verified: %s", err)
+	}
+
+	// Elapsed since the fault (i.e. since the higher of the two blocks)
+	faultAge := rt.CurrEpoch() - fault.Epoch
+	if faultAge <= 0 {
+		rt.Abortf(exitcode.ErrIllegalArgument, "invalid fault epoch %v ahead of current %v", fault.Epoch, rt.CurrEpoch())
+	}
+
+	// Note: only the first reporter of any fault is rewarded.
+	// Subsequent invocations fail because the target miner has been removed.
+	var st State
+	newlyVestedAmount := rt.State().Transaction(&st, func() interface{} {
+		newlyVestedFund, err := st.vestNewFunds(adt.AsStore(rt), rt.CurrEpoch())
+		AssertNoError(err)
+
+		return newlyVestedFund
+	}).(abi.TokenAmount)
+
+	totalLockedFund := st.LockedFunds
+	Assert(totalLockedFund.GreaterThanEqual(big.Zero()))
```

Comment:
Yes, there was a change in policy here (cc @davidad). We previously only burn the unvested balance and hence slasher share is a fraction of unvested balance. We have since moved to burning everything but the slasher share remains as a fraction of the unvested balance. Given that the slasher share will be subject to the same terms, im inclined to make slasher share a fraction of the total balance to be slashed instead of the unvested share.

Reply:
I am fine with slasher reward being a fraction of all burned tokens. In the event of premeditated hijinks, fully vested tokens would be already withdrawn by this point anyhow.

Reply:
I am fine with slasher reward being a fraction of all burned tokens. In the event of premeditated hijinks, fully vested tokens would be already withdrawn by this point anyhow.

Reply:
Changing to reward a fraction of all burnt tokens.